Feature/am2320 sensor

hieronymox opened the pull request:
I implemented a driver for the AM2320 sensor. The drawback of this sensor it only works to 100khz i2c clock. I couldn't find detailed info on how new features should be added to the codebase so I looked at other code in the repo.

I'm looking for feedback on both the changes and where I added them in the codebase :)

hieronymox committed:
AM2320 sensor support but only working with 100khz I2C

hieronymox committed:
Reverted back default i2c clock setting for sha2017 badge and added config setting to force the i2c clock to 100khz when using am2320

hieronymox committed:
fixed indent

hieronymox committed:
Added extra check for i2c clock speed in case someone manages to enable the am2320 with a higher clock

renzenicolai commented on firmware/components/driver_sensor_am2320/Kconfig:
```diff
@@ -0,0 +1,19 @@
+		select I2C_MASTER_FREQ_HZ
+
+	# This is used to force the already set clock value to 100khz
+	config I2C_MASTER_FREQ_HZ
```
Please do not force the I2C interface speed, let the user enter what they want: we can always error out on the existing CONFIG_I2C_MASTER_FREQ_HZ define of the I2C interface itself during compilation.

hieronymox replied:
Removed and added error macro to fail compile if wrong value is set

renzenicolai commented on firmware/components/driver_sensor_am2320/driver_am2320.c:
```diff
@@ -0,0 +1,155 @@
+    ESP_LOGD(TAG, "init called");
+
+    if (CONFIG_I2C_MASTER_FREQ_HZ > 100000) {
+        ESP_LOGE(
```
Maybe replace this with an error during compile time

#if CONFIG_I2C_MASTER_FREQ_HZ > 100000
#error "I2C interface speed is set to more than 100kHz, the AM2320 sensor supports a speed of at most 100kHz."
#endif

hieronymox replied:
Removed the if and added error macro to header file

renzenicolai commented on firmware/components/driver_sensor_am2320/driver_am2320.c:
```diff
@@ -0,0 +1,155 @@
+{
+    if (driver_am2320_init_done)
+        return ESP_OK;
+    ESP_LOGD(TAG, "driver is not enabled");
```
Please remove this debug log line.

hieronymox replied:
Removed together with other not useful code

renzenicolai commented on firmware/components/micropython/component.mk:
```diff
@@ -265,6 +265,11 @@ MP_EXTRA_INC += -I$(PROJECT_PATH)/components/libnmea/src/nmea
 MP_EXTRA_INC += -I$(PROJECT_PATH)/components/libnmea/src/parsers
 endif
 
+ifdef CONFIG_MICROPY_USE_AM2320
```
Just use your drivers CONFIG_DRIVER_AM2320_ENABLE here, don't add extra complexity if it's not usefull

hieronymox replied:
Changed to use the driver enable

renzenicolai commented on firmware/components/micropython/esp32/modam2320.c:
```diff
@@ -0,0 +1,52 @@
+STATIC mp_obj_t am2320_get_temperature() {
+    
+    float temperature = 0;
+    driver_am2320_get_temperature(&temperature);
```
What's the reason you're working with pointers instead of return values here?

hieronymox replied:
Good point that was left from some earlier code, changed to return value

renzenicolai reviewed (changes requested):
Thank you for putting in the effort of adding support for new hardware! Most of your change looks great but I do have a couple of small things that I'd like to see changed.

hieronymox committed:
Fixed PR comments:
- Use of return values for get_temperature and get_humidity
- Also use CONFIG_DRIVER_AM2320_ENABLE for python module
- Removed forced i2c clock setting from menuconfig for AM2320 module and change the check for I2C clock to compile time check
- Cleanup dummy init for AM2320 driver

hieronymox committed:
Use ESP error codes that match the type of error

hieronymox committed:
Added check if the driver is actually enabled before checking clock speed

Cat-Ion commented on firmware/components/driver_sensor_am2320/include/driver_am2320.h:
```diff
@@ -0,0 +1,32 @@
+#define SENSOR_NAN_VALUE 0xFFFF
+
+#ifdef CONFIG_DRIVER_AM2320_ENABLE
+	#if CONFIG_I2C_MASTER_FREQ_HZ > 100000
```
Please make sure you're not mixing tabs and spaces - the easiest way would be to just format everything using clang-format (there's a `.clang-format` file in the root of the repository)

hieronymox replied:
Yeah I must have done that manually, if I run format from my IDE (VSCode) it removes the indent but that seems to be inline with other code so I will commit that.

But regarding the clang-format I did see some codestyle that is not specified in the clang-format. For me to apply the same usage of braces and indent I have to use these extra settings: 

BreakBeforeBraces: Stroustrup
IndentWidth: 4

Is that the correct style? And if so shall I commit these changes to the .clang-format?

Cat-Ion replied:
The missing options should come from the parent style (Chromium). When I run `clang-format --style=file --dump-config` I see `BreakBeforeBraces: Attach` and `IndentWidth: 2`.

What do you mean when you say that you see missing settings? Did something warn you about them, or do you just see that they're not specified explicitly? If something (vscode?) warned you, can you check `clang-format --style=chromium --dump-config` with whatever clang-format binary it uses and check if the two options are in there? With clang-format 10 they're there, at least for me.

hieronymox replied:
Ahhh so 2 spaces indent and braces attached are what is intended. :) Yeah that is what I got initially. But I looked at a lot of code and noticed the formatting was not the same as what I got when I applied the formatting. So I assumed some settings where missing in the clang-format file.

Cat-Ion replied:
Yeah, the format file is rather recent, and we decided against formatting the whole repo at once. Instead we do a small formatting commit before touching any new file :)

hieronymox replied:
I will reformat my newly added files, but I wont reformat the change in i2c driver because the diff will be much bigger

hieronymox committed:
Fixed formatting

hieronymox committed:
Applied formatting

renzenicolai reviewed (approved):
lgtm
